test/test_transforms_v2_refactored.py

```diff
@@ -0,0 +1,326 @@
+        return default_tolerances
+
+    missing = default_tolerances.keys() - maybe_tolerance_dict.keys()
+    if missing:
+        raise pytest.UsageError("ADDME")
+
+    extra = maybe_tolerance_dict.keys() - default_tolerances.keys()
+    if extra:
+        raise pytest.UsageError("ADDME")
+
+    return maybe_tolerance_dict
+
+
+def _check_kernel_cuda_vs_cpu(kernel, tolerances, input, *args, **kwargs):
+    if input.device.type != "cuda":
+        return
+
+    input_cuda = input.as_subclass(torch.Tensor)
+    input_cpu = input_cuda.to("cpu")
+
+    actual = kernel(input_cuda, *args, **kwargs)
+    expected = kernel(input_cpu, *args, **kwargs)
+
+    assert_close(actual, expected, **tolerances, check_device=False)
+
+
+@cache
+def _script(fn):
+    try:
+        return torch.jit.script(fn)
+    except Exception as error:
+        raise AssertionError(f"Trying to `torch.jit.script` '{fn.__name__}' raised the error above.") from error
+
+
+def _check_kernel_scripted_vs_eager(kernel, tolerances, input, *args, **kwargs):
+    kernel_scripted = _script(kernel)
+
+    input = input.as_subclass(torch.Tensor)
+    actual = kernel_scripted(input, *args, **kwargs)
+    expected = kernel(input, *args, **kwargs)
+
+    assert_close(actual, expected, **tolerances)
+
+
+def _unbatch(batch, *, data_dims):
+    if isinstance(batch, torch.Tensor):
+        batched_tensor = batch
+        metadata = ()
+    else:
+        batched_tensor, *metadata = batch
+
+    if batched_tensor.ndim == data_dims:
+        return batch
+
+    return [
+        _unbatch(unbatched, data_dims=data_dims)
+        for unbatched in (
+            batched_tensor.unbind(0) if not metadata else [(t, *metadata) for t in batched_tensor.unbind(0)]
+        )
+    ]
+
+
+def _check_kernel_batched_vs_single(kernel, tolerances, input, *args, **kwargs):
+    input_type = datapoints.Image if is_simple_tensor(input) else type(input)
+    # This dictionary contains the number of rightmost dimensions that contain the actual data.
+    # Everything to the left is considered a batch dimension.
+    data_dims = {
+        datapoints.Image: 3,
+        datapoints.BoundingBox: 1,
+        # `Mask`'s are special in the sense that the data dimensions depend on the type of mask. For detection masks
+        # it is 3 `(*, N, H, W)`, but for segmentation masks it is 2 `(*, H, W)`. Since both a grouped under one
+        # type all kernels should also work without differentiating between the two. Thus, we go with 2 here as
+        # common ground.
+        datapoints.Mask: 2,
+        datapoints.Video: 4,
+    }.get(input_type)
+    if data_dims is None:
+        raise pytest.UsageError(
+            f"The number of data dimensions cannot be determined for input of type {input_type.__name__}."
+        ) from None
+    elif input.ndim <= data_dims or not all(input.shape[:-data_dims]):
+        # input is not batched or has a degenerate batch shape
+        return
```

Comment:
Two things here:

1. I've opted to perform the check whether an input is valid inside the `_check*` function rather than in the overarching `check` to keep the latter clean. This is not only the case here, but also in the other functions. Open for discussion on this one though.
2. This depends on https://github.com/pytorch/vision/pull/7562/files#r1202109254 so let's finish that discussion first.

Reply:
OK, did some testing and

> Does this happen

in fact it happens literally all the time and the rest of the test below is never hit. Probably because we only ever call `check_kernel` from the output of `make_image()` or `make_bounding_box()` which don't return batches.

Reply:
Yup. I haven't added it to the parametrization yet, but the `make_*` helpers have an option for adding batch dimensions:

https://github.com/pytorch/vision/blob/285500d6899b6aebac24718447c353c564838b31/test/common_utils.py#L497

By default, we are currently using these in the v2 tests:

https://github.com/pytorch/vision/blob/285500d6899b6aebac24718447c353c564838b31/test/common_utils.py#L429-L432

Reply:
> This depends on https://github.com/pytorch/vision/pull/7562/files#r1202109254 so let's finish that discussion first.

It seems my comment somehow got lost in the aether. I was saying that the current v2 tests include extra batch dimensions inside the regular parametrizations and thus this test needs to filter out the ones that are not applicable. We could also remove these from the regular parametrizations and perform a batching inside the test. 

That would make the parametrization simpler, but has the downside that only the batched vs unbatched test sees batched input at all. Meaning, if there was some issue with batched inputs on CUDA or JIT, we wouldn't know.

Reply:
I've opted for this in https://github.com/pytorch/vision/pull/7562#discussion_r1211320665.